src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -195,28 +200,23 @@ private void reconnect() {
      * Connects the websocket. Blocks until connected, or throws if the connection times out.
      */
     public void connect() throws ConnectionTimeoutException {
-        connect(true);
-    }
-
-    private void connect(boolean firstTrial) throws ConnectionTimeoutException {
         if (state == State.Connected) {
             log.warn("Trying to connect when already connected to " + options.getWebsocketApiUrl());
             return;
-        } else if (state == State.Connecting) {
+        } else if (state == State.Connecting || state == State.Sleeping) {
             log.warn("Trying to connect when already connecting to " + options.getWebsocketApiUrl());
-            waitForState(State.Connected);
+            waitForStateOrGiveUpAfterConnectionTimeout(State.Connected);
             return;
         }
-        state = State.Connecting;
 
         if (firstTrial) {
             log.info("Connecting to " + options.getWebsocketApiUrl() + "...");
-            websocket.connect();
-            waitForState(State.Connected);
+            firstTrial = false;
         } else {
             log.info("Reconnecting to " + options.getWebsocketApiUrl() + "...");
-            reconnect();
```

Comment:
Slight change: before we didn't block to wait for connection to be established. Now changed this method to always block and wait a bit if connecting / sleeping.